### PR TITLE
 create new CostSurface and CostSurfacePuData entities in dbs [MRXN23-93]

### PR DESCRIPTION
**Pull request by alexeh**
##### Creates CostSurface and CostSurfacePuData entities in dbs

Creates a new entity in each DB

NOTE: Commit tree clean-up pending.

### Overview

Adds a new table in each DB via 2 new migrations, along with it's entity representation.

### Testing instructions

Check that once you run the app, those new tables are created, and the required constraints are in place

### Feature relevant tickets

[MRXN23-93](https://vizzuality.atlassian.net/browse/MRXN23-93)
---

## Checklist before submitting

- [ ] Meaningful commits and code rebased on `develop`.
- [ ] If this PR adds feature that should be tested for regressions when
      deploying to staging/production, please add brief testing instructions
      to the deploy checklist (`docs/deployment-checklist.md`)
- [ ] Update CHANGELOG file

[MRXN23-93]: https://vizzuality.atlassian.net/browse/MRXN23-93?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: api/apps/api/test/utils/test-client/entity-seeder.ts [hunk 1]
```diff
@@ -0,0 +1,39 @@
+import { DataSource } from 'typeorm';
+import { apiConnections } from '@marxan-api/ormconfig';
+
+/**
+ * @description Utility function to create entities in the DB avoiding hitting the test API endpoints
```
Comment (hotzevzl) [inline api/apps/api/test/utils/test-client/entity-seeder.ts:5]: nice. With some "small" ergonomics improvements it could become a very useful (and kinda dangerously sharp) utility too.

Ergonomics improvements: I'd suggest mainly to avoid having to manually create dependencies and pass them on to the `seed()` call for dependent entities, which is a bit like juggling with sharp tools while hopping on a yoga ball.

But I do definitely see the place for something like this in use cases like the temporary test suite in this PR - well done!

File: api/apps/geoprocessing/src/app.module.ts [hunk 2]
```diff
@@ -10,7 +10,7 @@ import { PlanningUnitsModule } from '@marxan-geoprocessing/modules/planning-unit
 import { TileModule } from '@marxan-geoprocessing/modules/tile/tile.module';
 import { FeaturesModule } from '@marxan-geoprocessing/modules/features/features.module';
 import { ApiEventsModule } from '@marxan-geoprocessing/modules/api-events/api-events.module';
-import { SurfaceCostModule } from '@marxan-geoprocessing/modules/surface-cost/surface-cost.module';
+import { CostSurfaceModule } from '@marxan-geoprocessing/modules/cost-surface/cost-surface.module';
```
Comment (hotzevzl) [inline api/apps/geoprocessing/src/app.module.ts:13]: ❤️ for this change at last 😅 

Sometimes I still have nightmares with SurfaceCosts walking towards me like undead characters, muttering "surface-cost... surface-cost...". Not sure why I never unleashed my naming-things-properly OCD daemon back then. I think none of us was very sure about what the domain logic for all this would be, and naming was likewise confusing in our heads.

File: api/apps/geoprocessing/src/migrations/geoprocessing/1692254627921-AddsCostSurfacePuDataGeoDb.ts [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+  implements MigrationInterface {
+  public async up(queryRunner: QueryRunner): Promise<void> {
+    /**
+     * @todo: Check how cascade deletes should work
+     */
+    await queryRunner.query(`
+      CREATE TABLE cost_surface_pu_data (
+        id uuid PRIMARY KEY DEFAULT gen_random_uuid(),
+        cost_surface_id uuid NOT NULL,
+        puid uuid UNIQUE NOT NULL references projects_pu(id),
```
Comment (hotzevzl) [inline api/apps/geoprocessing/src/migrations/geoprocessing/1692254627921-AddsCostSurfacePuDataGeoDb.ts:13]: i think we can throw in an `on delete cascade`. By the time we remove `projects_pu` rows we would have already deleted the parent project, so we don't risk orphaning these `cost_surface_pu_data` rows

File: api/apps/api/src/migrations/api/1691980238543-AddCostSurdaceApiDb.ts [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+    CREATE TABLE cost_surfaces (
+      id uuid PRIMARY KEY DEFAULT gen_random_uuid(),
+      project_id uuid references projects(id),
+      name VARCHAR NOT NULL,
+      is_default BOOLEAN NOT NULL DEFAULT FALSE,
+      created_at timestamp without time zone NOT NULL DEFAULT now(),
+      created_by uuid NOT NULL REFERENCES users(id),
+      last_modified_at timestamp without time zone NOT NULL DEFAULT now(),
+      CHECK (name <> '')
+    );
```
Comment (hotzevzl) [inline api/apps/api/src/migrations/api/1691980238543-AddCostSurdaceApiDb.ts:15]: I suggest to also store here the `[min, max]` cost range of each cost surface.

it could be two float columns, or a 2-element float array (enforcing that cardinality is 2)

File: api/apps/api/src/modules/cost-surface/cost-surface-module-first.class.ts [hunk 5]
```diff
@@ -0,0 +1,8 @@
+import { Module } from '@nestjs/common';
+import { TypeOrmModule } from '@nestjs/typeorm';
+import { CostSurfaceApiEntity } from './cost-surface.api.entity';
+
+@Module({
+  imports: [TypeOrmModule.forFeature([CostSurfaceApiEntity])],
+})
+export class CostSurfaceModuleFirstClass {}
```
[the remark this answers is not in the document]
Comment (alexeh) [reply api/apps/api/src/modules/cost-surface/cost-surface-module-first.class.ts:8]: Yes, it is totally a tmp name, until, after sync-ing with @KevSanchez, we move everything related to cost surface at the top level. 
However, why `ProjectCostSurfaceModule` instead of `CostSurface` since we mentioned it to be a first class citizen?

File: api/apps/api/src/migrations/api/1691980238543-AddCostSurdaceApiDb.ts [hunk 6]
```diff
@@ -0,0 +1,31 @@
+import { MigrationInterface, QueryRunner } from 'typeorm';
+
+export class AddCostSurdaceApiDb1691980238543 implements MigrationInterface {
+  public async up(queryRunner: QueryRunner): Promise<void> {
+    await queryRunner.query(`
+    CREATE TABLE cost_surfaces (
+      id uuid PRIMARY KEY DEFAULT gen_random_uuid(),
+      project_id uuid references projects(id),
```
[the remark this answers is not in the document]
Comment (alexeh) [reply api/apps/api/src/migrations/api/1691980238543-AddCostSurdaceApiDb.ts:8]: Yeh, maybe best to talk about it later

File: api/apps/api/src/modules/projects/project.api.entity.ts [hunk 7]
```diff
@@ -153,6 +154,15 @@ export class Project extends TimeUserEntityMetadata {
...
 
+  @ApiPropertyOptional({ type: () => CostSurfaceApiEntity })
+  @OneToOne(
+    (_type) => CostSurfaceApiEntity,
+    (costSurface) => costSurface.project,
+  )
+  // TODO: if there will always be a (at least default) cost surface, this should be required, but it blows up the tests
+  //       double check if this is the case and refactor tests
+  costSurface?: CostSurfaceApiEntity;
+
```
[the remark this answers is not in the document]
Comment (alexeh) [reply api/apps/api/src/modules/projects/project.api.entity.ts:165]: Nice, thanks to clarifying, this is the sort of stuff I am still not sure about.

File: api/apps/geoprocessing/src/app.module.ts (hunk 2 above)
Comment (alexeh) [reply api/apps/geoprocessing/src/app.module.ts:13]: Alex The Reaper is happy making Andrea the Sensei happy

File: api/apps/geoprocessing/src/migrations/geoprocessing/1692254627921-AddsCostSurfacePuDataGeoDb.ts (hunk 3 above)
Comment (alexeh) [reply api/apps/geoprocessing/src/migrations/geoprocessing/1692254627921-AddsCostSurfacePuDataGeoDb.ts:13]: Nice clarification, thanks!

File: api/apps/api/src/migrations/api/1691980238543-AddCostSurdaceApiDb.ts (hunk 4 above)
Comment (alexeh) [reply api/apps/api/src/migrations/api/1691980238543-AddCostSurdaceApiDb.ts:15]: Again the sort of stuff I am not sure about. Let's bring this up later, cheers!

**Comment by alexeh** [comment]
Hey @hotzevzl thanks for the review, overall:|

1. Yes, there are changes made that are missing in this stage of the PR
2. All the naming is tmp until we bring everything to top level, which was decided to do in this PR after sync-ing with @KevSanchez 
3. There are domain-related suggested changes that I don't fully understand so:

Let's sync about it

File: api/apps/api/test/utils/test-client/entity-seeder.ts (hunk 1 above)
Comment (alexeh) [reply api/apps/api/test/utils/test-client/entity-seeder.ts:5]: Hm, I am not sure what you mean by ` avoid having to manually create dependencies and pass them`, some entities need to have some dependencies due to DB constraints (and some of them recursively).

If you mean encapsulating the creation of the dependencies to happen under the hood, I don't see how we can do that since this is a generic class. Unless we go for a utility for each entity.

And then, I didn't go for this option as I am still not sure how the data model works. Actually this was a great exercise to start understanding it!

File: api/apps/api/src/migrations/api/1691980238543-AddCostSurdaceApiDb.ts (hunk 6 above)
[the remark this answers is not in the document]
Comment (alexeh) [reply api/apps/api/src/migrations/api/1691980238543-AddCostSurdaceApiDb.ts]: addressed f460be51ecfc5c4457dab15240ba491640dc2e28

File: api/apps/geoprocessing/src/migrations/geoprocessing/1692254627921-AddsCostSurfacePuDataGeoDb.ts (hunk 3 above)
Comment (alexeh) [reply api/apps/geoprocessing/src/migrations/geoprocessing/1692254627921-AddsCostSurfacePuDataGeoDb.ts]: addressed f460be51ecfc5c4457dab15240ba491640dc2e28

File: api/apps/api/src/migrations/api/1691980238543-AddCostSurdaceApiDb.ts (hunk 6 above)
[the remark this answers is not in the document]
Comment (alexeh) [reply api/apps/api/src/migrations/api/1691980238543-AddCostSurdaceApiDb.ts]: addressed f460be51ecfc5c4457dab15240ba491640dc2e28

File: api/apps/api/src/modules/projects/project.api.entity.ts (hunk 7 above)
[the remark this answers is not in the document]
Comment (alexeh) [reply api/apps/api/src/modules/projects/project.api.entity.ts]: addressed f460be51ecfc5c4457dab15240ba491640dc2e28

File: api/apps/api/src/modules/users/user.api.entity.ts [hunk 8]
```diff
@@ -140,6 +141,12 @@ export class User {
 
   @OneToMany((_type) => IssuedAuthnToken, (token) => token.userId)
   issuedAuthnTokens?: IssuedAuthnToken[];
+
+  @OneToMany(
+    (_type) => CostSurfaceApiEntity,
+    (costSurface) => costSurface.createdBy,
+  )
+  costSurfaces?: CostSurfaceApiEntity[];
```
[the remark this answers is not in the document]
Comment (alexeh) [reply api/apps/api/src/modules/users/user.api.entity.ts:149]: addressed f460be51ecfc5c4457dab15240ba491640dc2e28

File: api/apps/api/src/modules/projects/project-cost-surface.controller.ts [hunk 9]
```diff
@@ -0,0 +1,85 @@
...
+    /**
+     * @todo: We will have to move ScenarioService.processCostSurfaceShapefile logic to another project-scoped
+     *        service and fire it from here.
+     */
+    throw new NotImplementedException();
+  }
+
+  @ImplementsAcl()
+  @UseGuards(JwtAuthGuard)
+  @Get(`:scenarioId/cost-surface`)
```
[the remark this answers is not in the document]
Comment (alexeh) [reply api/apps/api/src/modules/projects/project-cost-surface.controller.ts]: I believe this would meant breaking changes tho, lets clarify it later

File: api/apps/api/src/modules/projects/project-cost-surface.controller.ts [hunk 10]
```diff
@@ -0,0 +1,85 @@
...
+  constructor(
+    @Inject(forwardRef(() => ScenariosService))
+    public readonly scenarioService: ScenariosService,
+  ) {}
+
+  @ImplementsAcl()
+  @UseGuards(JwtAuthGuard)
+  @ApiOperation({ description: 'To be implemented' })
+  @ApiConsumesShapefile({ withGeoJsonResponse: false })
+  @GeometryFileInterceptor(GeometryKind.ComplexWithProperties)
```
[the remark this answers is not in the document]
Comment (alexeh) [reply api/apps/api/src/modules/projects/project-cost-surface.controller.ts]: Hm, this endpoint is part of moving cost-surface related operations from the scenarios controller to this one, I don't understand how / why the validations should change

image

File: api/apps/api/test/utils/test-client/entity-seeder.ts (hunk 1 above)
Comment (hotzevzl) [reply api/apps/api/test/utils/test-client/entity-seeder.ts:5]: yep, I was getting ahead of ourselves - I value the approach of validating db constraints this way, and if we need more of this kind of tests, we can find a way to make the approach feel more DSL-like and (to get, for example, a fully formed set of organization/user/project - for example, here - so that we can directly use the project when exercising the SUT, in this case the cost surface creation).

Basically instead of creating org/user/project in the tests before creating a cost surface, again following the example of these tests, we'd "just" create the project _through a utility function in the test itself_, and let the utility function encapsulate the creation of org and user needed for the project, using the same generic class under the hood.

File: api/libs/planning-unit-geometry/src/projects-pu.geo.entity.ts [hunk 11]
```diff
@@ -40,4 +43,15 @@ export class ProjectsPuEntity {
...
+
+  // @OneToMany(
+  //   () => CostSurfacePuDataGeoEntity,
+  //   (costSurfacePuData: CostSurfacePuDataGeoEntity) =>
+  //     costSurfacePuData.projectsPu,
+  //   { onDelete: 'CASCADE' },
+  // )
+  // costSurfacePuData!: CostSurfacePuDataGeoEntity[];
+  /**
+   * @todo: solve this doubt. move all geoprocessing cost surface stuff to libs? only the relevant entity?
```
Comment (hotzevzl) [inline api/libs/planning-unit-geometry/src/projects-pu.geo.entity.ts:55]: we didn't discuss this today - please let me know if you need more info on this, or even just a functional-design rubber ducky

**Review by hotzevzl — APPROVED** [review]
I may not have time to give another look at this once you finalise it, but you've addressed all the changes that were needed and you know what to do for the rest, following our sync this morning, so it's all basically pre-approved from my end 😄

File: api/apps/api/src/modules/projects/project-cost-surface.controller.ts (hunk 10 above)
[the remark this answers is not in the document]
Comment (hotzevzl) [reply api/apps/api/src/modules/projects/project-cost-surface.controller.ts]: yep, basically these enum values are "proxies" of typical limits we want to impose on shapefile file size, depending on... stuff. it's all a bit rule-of-thumb. in doubt, as Alicia!

File: api/apps/api/src/modules/projects/project.api.entity.ts (hunk 7 above)
[the remark this answers is not in the document]
Comment (hotzevzl) [reply api/apps/api/src/modules/projects/project.api.entity.ts]: > I recall mentioning that a default cost surface is created when a project is born, but not 100% sure, need to review the project creation flow

yep, we'll need to make this happen (it happens now in a different way, which is not adequate for id26.2)

[the remark this answers is not in the document]
Comment (hotzevzl) [reply api/apps/api/src/modules/projects/project.api.entity.ts]: tl;dr this should be an array of cost surfaces (which will be empty right at project creation time, and will then include at least the default cost surface for the project, once this has been created, and possibly more as users upload more)

File: api/libs/planning-unit-geometry/src/projects-pu.geo.entity.ts (hunk 11 above)
Comment (alexeh) [reply api/libs/planning-unit-geometry/src/projects-pu.geo.entity.ts:55]: Yep, I totally forgot to bring this up. Let's **extremely** quickly mention it in today's call.

File: api/apps/api/src/migrations/api/1691980238543-AddCostSurdaceApiDb.ts [hunk 12]
```diff
@@ -8,6 +8,8 @@ export class AddCostSurdaceApiDb1691980238543 implements MigrationInterface {
       project_id uuid NOT NULL references projects(id) ON DELETE CASCADE,
       scenario_id uuid references scenarios(id),
       name VARCHAR NOT NULL,
+      min float8,
+      max float8,
```
Comment (hotzevzl) [inline api/apps/api/src/migrations/api/1691980238543-AddCostSurdaceApiDb.ts:12]: depending on when we have this data in the overall flow, we may want to make these `NOT NULL` - though if we need to first create a cost surface and then only _at some later point_ we have min,max data then of course, ok as is

File: api/apps/api/src/modules/projects/project.api.entity.ts [hunk 13]
```diff
@@ -154,12 +154,9 @@ export class Project extends TimeUserEntityMetadata {
...
   scenarios?: Scenario[];
 
-  @ApiPropertyOptional({ type: () => CostSurfaceApiEntity })
-  @OneToOne(
-    (_type) => CostSurfaceApiEntity,
-    (costSurface) => costSurface.project,
-  )
-  costSurface?: CostSurfaceApiEntity;
+  @ApiPropertyOptional({ type: () => CostSurface })
+  @OneToOne((_type) => CostSurface, (costSurface) => costSurface.project)
```
Comment (hotzevzl) [inline api/apps/api/src/modules/projects/project.api.entity.ts:158]: please note, this should not be `@OneToOne()` - many cost surfaces can have the same `project_id`, and conversely here this would be `costSurface!: CostSurface[]` (which can very well be an empty array until we get to creating at least the default cost surface for a new project).

Comment (alexeh) [reply api/apps/api/src/modules/projects/project.api.entity.ts:158]: Yep, both this and above are unfinished (and untested) changes in a rush to wrap this up before the eagle caught me, but it was fast :(
